backends/bmv2/lower.cpp

```diff
@@ -127,34 +127,221 @@ const IR::Node* LowerExpressions::postorder(IR::Concat* expression) {
 
 /////////////////////////////////////////////////////////////
 
+namespace {
+
+struct VariableWriters {
+    std::set<const IR::AssignmentStatement*> writers;
+    VariableWriters() = default;
+    explicit VariableWriters(const IR::AssignmentStatement* writer)
+    { writers.emplace(writer); }
+    VariableWriters* join(const VariableWriters* other) const {
+        auto result = new VariableWriters();
+        result->writers = writers;
+        for (auto e : other->writers)
+            result->writers.emplace(e);
+        return result;
+    }
+    // Non-null only if there is exaclty one writer statement.
+    // Return the RHS of the assignment
+    const IR::Expression* substitution() const {
+        if (writers.size() != 1)
+            return nullptr;
+        auto first = *writers.begin();
+        return first->right;
+    }
+};
+
+struct VariableDefinitions {
+    std::map<cstring, const VariableWriters*> writers;
+    VariableDefinitions* clone() const {
+        auto result = new VariableDefinitions();
+        result->writers = writers;
+        return result;
+    }
+    VariableDefinitions* join(const VariableDefinitions* other) const {
+        auto result = clone();
+        for (auto e : other->writers)
+            result->writers.emplace(e);
+        return result;
```

Comment:
If `e.first` (the string) already has an entry in `writers`, then this won't actually do anything -- shouldn't it be unioning the sets in this case?  Or can it never happen (should probably add a BUG_CHECK if so)